cocos/2d/components/graphics.ts

```diff
@@ -738,7 +746,7 @@ export class Graphics extends UIRenderer {
                     for (let i = 0; i < renderDataList.length; i++) {
                         renderDataList[i].setRenderDrawInfoAttributes();
                     }
-                    const len = this.model!.subModels.length;
+                    const len = this._graphicsNativeProxy.getModel().subModels.length;
```

Comment:
Visiting `subModels` property of `Model` object really cost a lot.
In Model::getSubModels, it returns `const std::vector<cc::IntrusivePtr<cc::scene::SubModel>>&` , it will be converted to a new js array (which is a GC objec†) when visiting the getter.
```c++
static bool js_scene_Model_getSubModels(se::State& s) // NOLINT(readability-identifier-naming)
{
    auto* cobj = SE_THIS_OBJECT<cc::scene::Model>(s);
    // SE_PRECONDITION2(cobj, false, "Invalid Native Object");
    if (nullptr == cobj) return true;
    const auto& args = s.args();
    size_t argc = args.size();
    CC_UNUSED bool ok = true;
    if (argc == 0) {
        const std::vector<cc::IntrusivePtr<cc::scene::SubModel>>& result = cobj->getSubModels();
        ok &= nativevalue_to_se(result, s.rval(), nullptr /*ctx*/);
        SE_PRECONDITION2(ok, false, "Error processing arguments");
        SE_HOLD_RETURN_VALUE(result, s.thisObject(), s.rval());
        return true;
    }
    SE_REPORT_ERROR("wrong number of arguments: %d, was expecting %d", (int)argc, 0);
    return false;
}
SE_BIND_FUNC_AS_PROP_GET(js_scene_Model_getSubModels)

template <typename T>
inline bool nativevalue_to_se(const ccstd::vector<T> &from, se::Value &to, se::Object *ctx) { // NOLINT(readability-identifier-naming)
    se::HandleObject array(se::Object::createArrayObject(from.size()));
    se::Value tmp;
    for (size_t i = 0; i < from.size(); i++) {
        // If from[i] is on stack, then should create a new object, or
        // JS will hold a freed object.
        if CC_CONSTEXPR (!std::is_pointer<T>::value && is_jsb_object_v<T>) {
            auto *pFrom = ccnew T(from[i]);
            nativevalue_to_se(pFrom, tmp, ctx);
            tmp.toObject()->clearPrivateData();
            tmp.toObject()->setPrivateData(pFrom);
        } else {
            nativevalue_to_se(from[i], tmp, ctx);
        }

        array->setArrayElement(static_cast<uint32_t>(i), tmp);
    }
    to.setObject(array, true);
    return true;
}
```
So, if .subModels is visited frequently, I suggest don't invoke it in JS.

Reply:
Could it be done in CPP with the following `activeSubModel` & `updateData` ?